flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateFactory.java

```diff
@@ -0,0 +1,209 @@
+import org.apache.flink.metrics.Counter;
+import org.apache.flink.runtime.deployment.InputChannelDeploymentDescriptor;
+import org.apache.flink.runtime.deployment.InputGateDeploymentDescriptor;
+import org.apache.flink.runtime.deployment.ResultPartitionLocation;
+import org.apache.flink.runtime.io.network.ConnectionManager;
+import org.apache.flink.runtime.io.network.TaskEventPublisher;
+import org.apache.flink.runtime.io.network.buffer.BufferPool;
+import org.apache.flink.runtime.io.network.buffer.BufferPoolFactory;
+import org.apache.flink.runtime.io.network.buffer.NetworkBufferPool;
+import org.apache.flink.runtime.io.network.metrics.InputChannelMetrics;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionID;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionManager;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionType;
+import org.apache.flink.runtime.jobgraph.IntermediateDataSetID;
+import org.apache.flink.runtime.taskmanager.NetworkEnvironmentConfiguration;
+import org.apache.flink.runtime.taskmanager.TaskActions;
+import org.apache.flink.util.function.SupplierWithException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+
+import java.io.IOException;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Factory for {@link SingleInputGate} to use in {@link org.apache.flink.runtime.io.network.NetworkEnvironment}.
+ */
+public class SingleInputGateFactory {
+	private static final Logger LOG = LoggerFactory.getLogger(SingleInputGate.class);
+
+	private final boolean isCreditBased;
+
+	private final int partitionRequestInitialBackoff;
+
+	private final int partitionRequestMaxBackoff;
+
+	@Nonnull
+	private final ConnectionManager connectionManager;
+
+	@Nonnull
+	private final ResultPartitionManager partitionManager;
+
+	@Nonnull
+	private final TaskEventPublisher taskEventPublisher;
+
+	@Nonnull
+	private final NetworkBufferPool networkBufferPool;
+
+	private final int networkBuffersPerChannel;
+
+	private final int floatingNetworkBuffersPerGate;
+
+	public SingleInputGateFactory(
+		@Nonnull NetworkEnvironmentConfiguration networkConfig,
+		@Nonnull ConnectionManager connectionManager,
+		@Nonnull ResultPartitionManager partitionManager,
+		@Nonnull TaskEventPublisher taskEventPublisher,
+		@Nonnull NetworkBufferPool networkBufferPool) {
+
+		this.isCreditBased = networkConfig.isCreditBased();
+		this.partitionRequestInitialBackoff = networkConfig.partitionRequestInitialBackoff();
+		this.partitionRequestMaxBackoff = networkConfig.partitionRequestMaxBackoff();
+		this.connectionManager = connectionManager;
+		this.partitionManager = partitionManager;
+		this.taskEventPublisher = taskEventPublisher;
+		this.networkBufferPool = networkBufferPool;
+		this.networkBuffersPerChannel = networkConfig.networkBuffersPerChannel();
```

Comment:
It might be better to put `this.networkBuffersPerChannel` together with above `this.partitionRequestMaxBackoff`, because these infos are all form `networkConfig`.